crates/api/src/space.rs

```diff
@@ -0,0 +1,82 @@
+//! Kitsune2 space related types.
+
+use crate::*;
+use std::sync::Arc;
+
+/// Handler for events coming out of Kitsune2 such as messages from peers.
+pub trait SpaceHandler: 'static + Send + Sync + std::fmt::Debug {
+    /// We have received an incomming message from a remote peer.
+    ///
+    /// If this callback handler returns an `Err` response, the connection
+    /// will be closed immediately.
+    //
+    // Note: this is the minimal low-level messaging unit. We can decide
+    //       later if we want to handle request/response tracking in
+    //       kitsune itself as a convenience or if users of this lib
+    //       should have to implement that if they want it.
+    fn incoming_message(
+        &self,
+        peer: AgentId,
+        data: bytes::Bytes,
+    ) -> K2Result<()>;
+}
+
+/// Trait-object [SpaceHandler].
+pub type DynSpaceHandler = Arc<dyn SpaceHandler>;
+
+/// Represents a unique dht space within which to communicate with peers.
+pub trait Space: 'static + Send + Sync + std::fmt::Debug {
+    /// Get a reference to the peer store being used by this space.
+    /// This could allow you to inject peer info from some source other
+    /// than gossip or bootstrap, or to query the store directly if
+    /// you have a need to determine peers for direct messaging.
+    fn peer_store(&self) -> &peer_store::DynPeerStore;
+
+    /// Indicate that an agent is now online, and should begin receiving
+    /// messages and exchanging dht information.
+    fn local_agent_join(
```

Comment:
Why are these methods for local agents? Is it because you can online join a particular kitsune node if you're local?

Reply:
The distinguishing feature of a "Local" agent vs a generic "Agent" or "AgentInfo" is that a local agent is able to sign messages. So this function would be used in conductor when a cell is enabled, it provides an instance of a "LocalAgent" to kitsune that is able to sign agent infos and make that peer discoverable online.